### PR TITLE
74 declare side effects

### LINKED ISSUE #74: Allow steps to declare if they change external state or not
eg if you have a workflow which pulls data, and then does 5 filters on it, then we would consider that a valid workflow. But it doesn't actually do anything. We want to be able to say have a step which updates something, or sends an email etc. These steps would be declared as changing state. A valid workflow must have all finishing steps in the graph change state?? Maybe this is a warning not an error, we will want to run half authored workflows which might not do anything yet.

* [ ] This can apply to flow steps or connector steps
* [ ] show these in the table, maybe something like a lightning icon
* [ ] Show these in the graph visually somehow, maybe a color or dotted border (will also depend on if it is a connector or flow step)

### PR DESCRIPTION
**Pull request by keevan**
- fix: issue if exception is thrown and engine is not defined to log it
- feat: add declaration of side effects and checks
- feat: add basic icon outputs to the different types of steps
- feat: add different styles for different step types in the visual graph

Resolves #74 
Related to #107 - still needs arrow styles

### CODE REVIEW

**Comment by keevan** [comment]
image

Example flow:
[newww_20220602_0706.yml.txt](https://github.com/catalyst/moodle-tool_dataflows/files/8821503/newww_20220602_0706.yml.txt)

**Comment by keevan** [comment]
After most recent push:

image

image

File: classes/dataflow.php [hunk 1]
```diff
@@ -313,38 +314,17 @@ public function validate_dataflow() {
     /**
      * Returns a dotscript of the dataflow, false if no connections are available
      *
-     * @return     string|false dotscript or false if not a valid flow
+     * @return     string dotscript or false if not a valid flow
```
Comment (jaypha) [inline classes/dataflow.php:317]: This needs to be updated if the function no longer returns false.

Comment (keevan) [reply classes/dataflow.php:317]: Ah nice catch. I'll fix this up.

**Pull request merged**